ldrshared/clients/messenger/pubsub.py

```diff
@@ -0,0 +1,250 @@
+
+    def _housekeeping(self):
+        """
+        Simple mechanism for stopping the inline database from getting
+        too bloated by dropping records that are older than:
+
+        datetime.now() - self.message_retention # a datetime.timedelta()
+        """
+        self.db: TinyDB
+
+        # Note: .doc_id's are auto incrementing.
+        record_id = min([int(x.doc_id) for x in self.db])
+        passes = 0
+        max_passes = len(self.db)
+
+        while True:
+            record = self.db.get(doc_id=record_id)
+            time_stamp_as_datetime = datetime.datetime.strptime(
+                record["time_stamp"], STR_TIME
+            )
+            if time_stamp_as_datetime <= (
+                datetime.datetime.now() - self.message_retention
+            ):
+                logging.warning(f"Length was {len(self.db)}")
+                self.db.remove(doc_ids=[record_id])
+                logging.warning(f"Length is {len(self.db)}")
+                record_id = str(int(record_id) + 1)
+            else:
+                logging.warning(
+                    f"Comparing: {time_stamp_as_datetime} and {datetime.datetime.now() - self.message_retention}"
+                )
+                break
+
+            # Never look beyond the last populated index
+            passes += 1
+            if passes == max_passes:
+                break
+
+
+class PubSubClient(BaseMessenger):
+    def _setup(self, dbname="db.json", message_retention={"days": 7}):
+        """
+        Setup for pubsub client
+
+        dbname: Filename for the Tinydb inline message database,
+                configurable for testing.
+        message_retention: Allows altering of retention period
+                for inline message database, configurable for testing,
+        """
+        self._project_id = GCP_PROJECT_ID
+        self.message_buffer = []
+        self.deduplicator = Deduplicator(dbname, message_retention)
+        self.found_message_duplicate = False
+
+    def subscribe(self, subscription: pubsub_gapic_types.Subscription):
+        """
+        Subscribe the messenger client
+        """
+        subscriber_client = SubscriberClient()
+
+        subscription_path: str = subscriber_client.subscription_path(
+            GCP_PROJECT_ID, subscription.name.split("/")[-1]
+        )
+
+        def callback(message: Message) -> None:
+            # Is message in database as previously processed?
+            if self.deduplicator.is_new_message(message):
+                message.nack()
+                # Is message already in buffer waiting to be processed?
+                if not self.deduplicator.message_is_in_list(
+                    message, self.message_buffer
+                ):
+                    self.message_buffer.append(message)
+            else:
+                message.ack()
+                self.found_message_duplicate = True
+
+        self.streaming_pull_future: StreamingPullFuture = subscriber_client.subscribe(
```

Comment:
I'm hoping this `callback` functionality doesn't try to do any kind of parallelism?

Reply:
quite the opposite, I cant remove the nasty futures bit as its baked into PubSub (its very javascript promises. esque..ewww) but the client is flattening all that into a much simple sequential model.